### ISSUE TITLE
question regarding how losses are combined

### ISSUE THREAD

**Issue by foobar8675** [pytorch/xla#2705]
i added logging of the gen loss to the DC-GAN example `print(f"{device} {g_error.item()}")`
https://colab.research.google.com/drive/1lUPh7bljOw84jMYGX-UcYYXvqqzW895Q?usp=sharing

and I frequently see the losses for multiple devices to be the same, ie: 
xla:0 0.6018649935722351
xla:1 0.6018649935722351

in a case like this, is xl1:0 and xla:1 doing the same exact calculation? I'm not sure how that could be since they each have a different batch right? also, can someone point me to a link that explains at a high level, how the different cores all get combined?

**Comment by taylanbil** [comment]
I can't see the colab as I need access, but I'm assuming you're all-reducing the losses? so they end up the same.

Generally, in a data-parallel program (most common usecase), you would have all the cores do the **same** work on a **different** batch. And thus obtain different losses. The gradients are then all-reduced so model weights are same across cores. Then when you report loss, you may sync them by doing an all-reduce before reporting, so you report just one loss having information across all devices.

**Comment by taylanbil** [comment]
[here](https://github.com/pytorch/xla/blob/master/test/test_train_mp_imagenet.py#L208) is where the gradient all-reduce happens, as part of [`xm.optimizer_step`](https://github.com/pytorch/xla/blob/master/torch_xla/core/xla_model.py#L780)

[here](https://github.com/pytorch/xla/blob/master/test/test_train_mp_imagenet.py#L228) is an example of an all-reduce of accuracy value.

**Comment by foobar8675** [comment]
I'm so sorry about the link @taylanbil .  i updated the link above.  

just to double check my understanding, when `xm.optimizer_step` is called, it updates all the gradients (or reduces) from all 8 cores to the master core?  in the case of the `DC-GAN` (which was copied directly from the pytorch/xla github readme), I am seeing the same loss on different cores which is what is so confusing.

```
ordinal:[0] 0.6541286706924438
ordinal:[5] 0.6541286706924438
```

**Comment by taylanbil** [comment]
`xm.optimizer_step` averages all the gradients in all cores. so all cores have the same exact gradients. otherwise your model weights would diverge, and essentially it would be training different models on each core. This is standard in all data-parallel distributed setups, not torch_xla specific at all.

If you don't allreduce the losses, then the loss values should be different on cores. I'm assuming you're reducing before reporting. Otherwise, something is wrong. One way this would happen is, if you don't shard the data, and essentially each core is processing the same batch, so you're getting the dame output because cores holding the same model and having the same input.

**Comment by foobar8675** [comment]
got it.

and yes, (i didn't author the example code, but should have read it more carefully) `xm.optimizer_step` is called before logging, so that does make sense. thanks for explaining!
